File: GEOSmkiau_GridComp/pyMKIAU/interface/interface.c
#include <stdio.h>
#include <time.h>
#include "interface.h"

extern int pyMKIAU_interface_c_setservice()
{
    // Check magic number
    int return_code = pyMKIAU_interface_py_setservices();

    if (return_code < 0)
    {
        exit(return_code);
    }
}

extern int pyMKIAU_interface_c_run(a_pod_struct_t *options, const float *in_buffer, float *out_buffer)
{
    // Check magic number
    if (options->mn_123456789 != 123456789)
    {
        printf("Magic number failed, pyMKIAU interface is broken on the C side\n");
        exit(-1);
    }

    int return_code = pyMKIAU_interface_py_run(options, in_buffer, out_buffer);

    if (return_code < 0)
    {
        exit(return_code);
    }
}

File: GEOSmkiau_GridComp/pyMKIAU/interface/interface.f90
module pyMKIAU_interface_mod

   use iso_c_binding, only: c_int, c_float, c_double, c_bool, c_ptr

   implicit none

   private
   public :: pyMKIAU_interface_f_setservice, pyMKIAU_interface_f_run
   public :: a_pod_struct_type

   !-----------------------------------------------------------------------
   ! See `interface.h` for explanation of the POD-strict struct
   !-----------------------------------------------------------------------
   type, bind(c) :: a_pod_struct_type
      integer(kind=c_int) :: npx
      integer(kind=c_int) :: npy
      integer(kind=c_int) :: npz
      ! Magic number
      integer(kind=c_int) :: make_flags_C_interop = 123456789
   end type


   interface

      subroutine pyMKIAU_interface_f_setservice() bind(c, name='pyMKIAU_interface_c_setservice')
      end subroutine pyMKIAU_interface_f_setservice

      subroutine pyMKIAU_interface_f_run(options, in_buffer, out_buffer) bind(c, name='pyMKIAU_interface_c_run')

         import c_float, a_pod_struct_type

         implicit none
         ! This is an interface to a C function, the intent ARE NOT enforced
         ! by the compiler. Consider them developer hints
         type(a_pod_struct_type), intent(in) :: options
         real(kind=c_float), dimension(*), intent(in) :: in_buffer
         real(kind=c_float), dimension(*), intent(out) :: out_buffer

      end subroutine pyMKIAU_interface_f_run
   
   end interface

end module pyMKIAU_interface_mod
